### ISSUE TITLE
monitor the GPU memory, log the visual's and UI frame's life cycle

### ISSUE THREAD

**Issue by yxdragon** [vispy/vispy#2083]
```python
from vispy import scene
from vispy.geometry import create_box

canvas = scene.SceneCanvas(keys='interactive', bgcolor='white', size=(800, 600), show=True)
view = canvas.central_widget.add_view()

vts, faces, ls = create_box(1,2,3, 1,2,3)
box = scene.visuals.Mesh(vertices=vts['position'], faces=faces, shading='smooth', parent=view.scene)
box.parent = None
del box
view.camera = scene.TurntableCamera(elevation=30, azimuth=30)
```
add __del__ method in Mesh visual class.
I add a mesh, then set the parent None, and del it. but it is not deleted, until I close the window.
is it a bug? or some other way to release the mesh object?

I think we need to do some work to monitor the gpu memory usage, log how many mesh are alive, when to release. 
And also every UI frame work, to make sure the frame, panel, canvas is released after close.

May be it is a hard work, but at least monitor and found these bugs first.

**Comment by djhoese** [comment]
You brought some of this up in #2065. Some of this may be specific to the Wx backend, we'd have to check.

As far as I can tell the Visuals as you've used them above, should be deleted properly and garbage collected. However, since Python will garbage collect whenever it wants to, it may be hard to know when that will happen.

**Comment by yxdragon** [comment]
```python
from vispy import scene
from vispy.geometry import create_box
import sys, weakref

canvas = scene.SceneCanvas(keys='interactive', bgcolor='white', size=(800, 600), show=True)
view = canvas.central_widget.add_view()

vts, faces, ls = create_box(1,2,3, 1,2,3)
box = scene.visuals.Mesh(vertices=vts['position'], faces=faces, shading='smooth', parent=view.scene)

weakbox = weakref.ref(box)
print(sys.getrefcount(weakbox()))

box.parent = None
weakbox = weakref.ref(box)
print(sys.getrefcount(weakbox()))

del box
print(sys.getrefcount(weakbox()))

view.camera = scene.TurntableCamera(elevation=30, azimuth=30)


if __name__ == '__main__':
    canvas.app.run()
```

I remove the del overwrite, for I saw some blog say "overwrite del would affect the gc", then use sys.getrefcount. but after many test, I can not trust the visuals has been deleted.

And the ui backend's leak, I had tracked the bug, and found the matter. I think it is not specific for wx, but for all UI framework. (maybe exclude the web backend)

I re mention it here, because I think, vispy do many work and test for various function, but few test for the base framework's stability and safety. I think it is important for industry application, So I suggest to do some work, at least monitor it firstly.

**Comment by djhoese** [comment]
I agree that this type of stuff is important, but I can't guarantee that any of the core developers have the time (or need) to do an in-depth look at this issue. It is also difficult when you haven't provided expected output from your test code above. I'm also not completely sure what the expected results from every step of this should be. I've modified your script a little so I could explore this a little further:

```python
from vispy import scene
from vispy.geometry import create_box
from vispy import visuals
import sys, weakref
import gc

canvas = scene.SceneCanvas(keys='interactive', bgcolor='white', size=(800, 600), show=True)
view = canvas.central_widget.add_view()

vts, faces, ls = create_box(1,2,3, 1,2,3)
box = scene.visuals.Mesh(vertices=vts['position'], faces=faces, shading='smooth', parent=view.scene)
#box = scene.visuals.Mesh(vertices=vts['position'], faces=faces, shading='smooth')
#box = visuals.MeshVisual(vertices=vts['position'], faces=faces, shading='smooth')
print("Ref count after creation: ", sys.getrefcount(box))

weakbox = weakref.ref(box)
gc.collect()
print("Ref count after weakref creation: ", sys.getrefcount(box))

if isinstance(box, scene.visuals.Mesh) and box.parent is not None:
    box.parent = None
    gc.collect()
    print("Ref count after unsetting parent: ", sys.getrefcount(box))

del view
gc.collect()
print("Weak ref count after view delete:", sys.getrefcount(box))

del canvas
gc.collect()
print("Weak ref count after canvas delete:", sys.getrefcount(box))

del box
gc.collect()
if weakbox() is not None:
    print("Weak ref count after box delete:", sys.getrefcount(weakbox()))
```

Near the top I have three options for creating the Mesh:

1. With a parent assigned
2. With a parent never assigned
3. Using the lower-level MeshVisual class which doesn't involve the SceneCanvas Node class at all

I print out the reference counting where I can and get these results for each case.

### Parent assigned at creation

Ref count after creation:  7
Ref count after weakref creation:  7
Ref count after unsetting parent:  6
Ref count after view delete: 6
Ref count after canvas delete: 6

### No parent assigned

Ref count after creation:  5
Ref count after weakref creation:  5
Ref count after view delete: 5
Ref count after canvas delete: 5

### MeshVisual

Ref count after creation:  3
Ref count after weakref creation:  3
Ref count after view delete: 3
Ref count after canvas delete: 3

**Disclaimer**: I could be doing this completely wrong. I think I sometimes know what I'm doing.

So while we see the total number of reference counts changes with the creation case, it does go down when we set the parent to `None`. This means we're at least doing it a little correctly. I should also point out that the parent property on the visual already uses weakrefs:

https://github.com/vispy/vispy/blob/acafcb07a0c9947c319e1611d72cf7045670a8e7/vispy/scene/node.py#L199

I'll try a few other cases and see if I can get a better understanding of where these references are coming from.

@yxdragon You've talked about how things aren't deleted when you delete them and other memory issues and stuff, but how are you actually monitoring this? How do you know it isn't getting deleted? Are you worried only about GPU memory or are we worried about CPU memory too? It is very possible that we aren't explicitly deleting GPU resources when a Python object is getting destroyed. That could be something we consider for the future.

**Comment by djhoese** [comment]
I just added this:

```python
if hasattr(box, "parent") and box.parent is None:
    box.parent = view.scene
    gc.collect()
    print("Ref count after setting parent: ", sys.getrefcount(box))
```

after the weakref creation and this results in:

```
Ref count after creation:  5
Ref count after weakref creation:  5
Ref count after setting parent:  7
Ref count after unsetting parent:  6
Ref count after view delete: 6
Ref count after canvas delete: 6
```

This tells me that the parent/node stuff is creating two references to the object and only getting rid of one. CC @almarklein in case he has any ideas.

**Comment by djhoese** [comment]
See #2089 for a solution for getting the reference counts back to an expected state. Not sure it is a "solution" to anything though.
